File: frontend/src/api/insights.ts
import apiClient from "../lib/apiClient";
import type { Employee } from "../types/employee";
import type {
  CountryInsight,
  DepartmentInsight,
  DistributionBucket,
  JobTitleInsight,
  MetaFilters,
  SummaryStats,
} from "../types/insights";

export async function getSummary(): Promise<SummaryStats> {
  const { data } = await apiClient.get<SummaryStats>("/insights/summary");
  return data;
}

export async function getByCountry(country?: string): Promise<{ data: CountryInsight[] }> {
  const { data } = await apiClient.get<{ data: CountryInsight[] }>("/insights/by-country", {
    params: { country: country ?? "" },
  });
  return data;
}

export async function getByJobTitle(country?: string, jobTitle?: string): Promise<{ data: JobTitleInsight[] }> {
  const { data } = await apiClient.get<{ data: JobTitleInsight[] }>("/insights/by-job-title", {
    params: { country: country ?? "", job_title: jobTitle ?? "" },
  });
  return data;
}

export async function getByDepartment(): Promise<{ data: DepartmentInsight[] }> {
  const { data } = await apiClient.get<{ data: DepartmentInsight[] }>("/insights/by-department");
  return data;
}

export async function getSalaryDistribution(
  country?: string,
  jobTitle?: string,
): Promise<{ buckets: DistributionBucket[] }> {
  const { data } = await apiClient.get<{ buckets: DistributionBucket[] }>("/insights/salary-distribution", {
    params: { country: country ?? "", job_title: jobTitle ?? "" },
  });
  return data;
}

export async function getTopPaid(
  limit = 10,
  country?: string,
  department?: string,
): Promise<{ data: Employee[] }> {
  const { data } = await apiClient.get<{ data: Employee[] }>("/insights/top-paid", {
    params: { limit, country: country ?? "", department: department ?? "" },
  });
  return data;
}

export async function getMetaFilters(): Promise<MetaFilters> {
  const { data } = await apiClient.get<MetaFilters>("/meta/filters");
  return data;
}

File: backend/app/routers/insights.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.services.insights_service import (
    get_by_country,
    get_by_department,
    get_by_job_title,
    get_salary_distribution,
    get_summary,
    get_top_paid,
)

router = APIRouter(prefix="/insights", tags=["insights"])


@router.get("/summary")
def summary_endpoint(db: Session = Depends(get_db)):
    return get_summary(db)


@router.get("/by-country")
def by_country_endpoint(country: str | None = None, db: Session = Depends(get_db)):
    return get_by_country(db, country)


@router.get("/by-job-title")
def by_job_title_endpoint(
    country: str | None = None,
    job_title: str | None = None,
    db: Session = Depends(get_db),
):
    return get_by_job_title(db, country, job_title)


@router.get("/by-department")
def by_department_endpoint(db: Session = Depends(get_db)):
    return get_by_department(db)


@router.get("/salary-distribution")
def salary_distribution_endpoint(
    country: str | None = None,
    job_title: str | None = None,
    db: Session = Depends(get_db),
):
    return get_salary_distribution(db, country, job_title)


@router.get("/top-paid")
def top_paid_endpoint(
    limit: int = Query(default=10, ge=1),
    country: str | None = None,
    department: str | None = None,
    db: Session = Depends(get_db),
):
    return get_top_paid(db, limit, country, department)
